core/sql/parser/sqlparser.y

```diff
@@ -6579,6 +6581,9 @@ table_reference : table_name_and_hint
               | rel_subquery_and_as_clause
                 { $$ = $1; }
 
+              | with_clause_list
```

Comment:
The ANSI SQL standard defines the "with" clause as part of a query_expression, so I would recommend doing the same here (non_join_query_expression in our case). Also, it would be good to accept the WITH RECURSIVE syntax and to give an error that this is not yet supported.